### DISCUSSION TITLE
`List[...]` as type for a `Form` field doesn't work as expected

### DISCUSSION THREAD

**Discussion by HitLuca** [fastapi/fastapi#8741] [Questions]
Opening a new issue since no action has been taken on https://github.com/tiangolo/fastapi/issues/842 for more than a month.

Looks like we have an issue on how fastapi handles lists of elements passed as `Form` parameter (so anything that is type hinted `List[...] = Form(...)`). The issue doesn't arise when they are defined as `Query` parameters.
After some digging I managed to get to the root of the problem, which is _how_ we pass the parameters when making an api call. fastapi indeed correctly understands that the input needs to be a list of elements, but doesn't parse the list if it comes from a single field.

Example using `Query`, works as  expected

```python
@app.post("/")
def api_test(l: List[int] = Query(...)) -> List[int]:
    return l
```
Api request generated using the swagger UI
```bash
curl -X 'GET' \
  'http://localhost:5009/?l=1&l=2' \
  -H 'accept: application/json'
```
note how the passed values are `l=1&l=2`
```bash
[
  1,
  2
]
```

Example using `Form`, doesn't work as  expected

```python
@app.post("/")
def api_test(l: List[int] = Form(...)) -> List[int]:
    return l
```
Api request generated using the swagger UI
```bash
curl -X 'POST' \
  'http://localhost:5009/' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'l=1,2'
```
note how the passed values are `l=1,2`
```bash
{
  "detail": [
    {
      "loc": [
        "body",
        "l",
        0
      ],
      "msg": "value is not a valid integer",
      "type": "type_error.integer"
    }
  ]
}
```

Manual api request
```bash
curl -X 'POST' \
  'http://localhost:5009/' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'l=1' \
  -d 'l=2'
```
Now the values are passed as `-d 'l=1'` and `-d 'l=2'`
```bash
[
  1,
  2
]
```

I am fairly sure that `l=1,2` should be an accepted way of passing a list of values as parameter, but fastapi doesn't seem to like it. Also, if this isn't the case, the Swagger UI doesn't produce the correct curl requests for lists given as `Form` parameters.

Packages:
- `fastapi`: `0.66.0`
- `python`: `3.7.4`

Let me know if you need other details!

Thank you for coming to my TED talk

**Comment by uricholiveira** [comment]
For the media type application/x-www-form-urlencoded, you need to use like this:

![Captura de tela de 2021-07-19 05-02-41](https://user-images.githubusercontent.com/48108280/126124770-f84a153c-5d4f-4390-bd82-0a554e3eac5f.png)

`l=1,2` is only accepted by query params in GET requests. If you doens't like the way above, you can use JSON.

**Comment by HitLuca** [comment]
Then why does the Swagger UI produce curl requests which clearly don't work?

**Comment by uricholiveira** [comment]
You've got me! I don't know what is causing it. 😢

**Comment by HitLuca** [comment]
Found a new related issue, when using `pydantic` `BaseModel` as `Form` parameters:

```
class Color(BaseModel):
    r: float
    g: float
    b: float

@app.post("/")
def api_test(color: Color = Form(...)):
    print(color)
    print(type(color))
```

Request made using the Swagger UI:

```
curl -X 'POST' \
  'http://localhost:5010/' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'color={
  "r": 0,
  "g": 1,
  "b": 2
}'
```

Response:

```
{
  "detail": [
    {
      "loc": [
        "body",
        "color"
      ],
      "msg": "value is not a valid dict",
      "type": "type_error.dict"
    }
  ]
}
```

Again: is the implementation wrong? Is Swagger wrong? If the implementation is bad why isn't it catched as an issue?

Would really like to get some kind of feedback on this or even an acknowledgement that somebody is remotely planning to look into it, seems like nobody cares (@tiangolo ?)

**Reply by jagadeeshkoka** [reply]
@HitLuca  - is this issue fixed ? I am having exactly same issue

**Comment by HitLuca** [comment]
Bump

**Comment by MarioIshac** [comment]
This is also a problem with `Tuple`, not just `List`. Additionally, now in FastAPI 0.68.0 (but not 0.67.0), the Swagger UI doesn't load for endpoints with `Form` parameters of type `Tuple`. See #3665.

**Comment by francoposa** [comment]
Bump on this, still experiencing this with lists of form params

**Comment by StevRamos** [comment]
Why hasn't it been solved? I am experiencing it too. @tiangolo any comment about this? Thank you in advance

The service:
```python
@router.post(
    path="/solicitud-inscripcion/registrar",
    tags=["identificacion"]
)
async def registrar_solicitud_inscripcion(
    imagenes: UploadFile = File(...),
    esParcial: bool = Form(
        ...,
        example=True
    ),
    idSolicitante: int = Form(
        ...,
        gt=0,
        example=1
    ),
    idSede: int = Form(
        ...,
        gt=0,
        example=1
    ),
    idPropietario: int = Form(
        ...,
        gt=0,
        example=1
    ),
    bienes: List[BienOutModel] = Form(...)
):
    return {"resp": "it works"}
```

```
curl -X 'POST' \
  'http://127.0.0.1:8000/identificacion/solicitud-inscripcion/registrar' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'imagenes=@16.pdf;type=application/pdf' \
  -F 'esParcial=false' \
  -F 'idSolicitante=1' \
  -F 'idSede=1' \
  -F 'idPropietario=1' \
  -F 'bienes={
  "id": 1,
  "nombre": "Ramos",
  "descripcion": "Ramos",
  "sumilla": "Ramos",
  "imagenes": [
    {
      "id": 1,
      "nombre": "Ramos",
      "url": "Ramosurl",
      "idOrientacion": 1,
      "idBien": 1,
      "activo": false
    }
  ]
}'
```

and the output is this

```
{
  "detail": [
    {
      "loc": [
        "body",
        "bienes",
        0
      ],
      "msg": "value is not a valid dict",
      "type": "type_error.dict"
    }
  ]
}
```

**Reply by jagadeeshkoka** [reply]
Is it solved in FastAPI ? @StevRamos , I am also facing the same issue

**Comment by phillipuniverse** [comment]
> Then why does the Swagger UI produce curl requests which clearly don't work?

> You've got me! I don't know what is causing it. 😢

~@HitLuca @uricholiveira probably the miss here is that FastAPI/Pydantic need to use the [`explode` keyword in the OpenAPI specification](https://swagger.io/docs/specification/serialization/). It looks like the default is `simple` in the OpenAPI spec which yields the CSV form.~


EDIT - this is a bug in swagger-ui see https://github.com/swagger-api/swagger-js/issues/2713

That linked doc talks about query parameters but there is a mention of it [when describing request bodies too](https://swagger.io/docs/specification/describing-request-body/). Looks like the same rules apply.

**Comment by falkben** [comment]
@phillipuniverse Do you override the OpenAPI generation to fix this? Any chance you'd submit a PR to fix this here or in pydantic?

**Comment by falkben** [comment]
I'd like to do a bit more with this, but it's getting late here, and wanted to put something down because I did make a tiny bit of progress with this and wanted to share. 

[This page](https://swagger.io/docs/specification/describing-request-body/) describes where the `explode` needs to go in the OpenAPI file

It goes under the content / multipart-form (or x-www-form-urlencoded) section in it's own `encoding` section ("beside" schema).

@phillipuniverse is correct -- once adding the `explode: true` configuration to the openapi schema, swaggerUI correctly produces requests into the API which don't treat the param as just a singular string but as a list of values.

As an example of what the openapi schema should look like ... if you had a param that was a form list (`my_list_param: List[str] = Form(...)`) the openapi.json to get the swaggerUI to work correctly with it would be:

```jsonc
"requestBody": {
  "content": {
    "multipart/form-data": {
      "schema": {
        "$ref": "#/components/schemas/Body_update_movie_movie__movie_id__patch"
      },
      // this is the section to add:
      "encoding": {
        "my_list_param": {
          "style": "form",
          "explode": true
        }
      }
    }
  }
}
```

**Comment by phillipuniverse** [comment]
@falkben looking at this again, I think that there is a bug in Swagger UI.  According to the 'Form Data' section of [describing request bodies](https://swagger.io/docs/specification/describing-request-body/), it says:

> By default, arrays are serialized as array_name=value1&array_name=value2 and objects as prop1=value1&prop=value2, but you can use other serialization strategies as defined by the OpenAPI 3.0 Specification

So the default should work the exact same as query parameters. If there are changes to code, it should be within Swagger UI, not in FastAPI or Pydantic (although perhaps FastAPI could make it easier to extend the `encoding`).

I messed around with a very rough way to override this with monkeypatching:

```python
    import fastapi.openapi.utils
    orig_get_request_body = fastapi.openapi.utils.get_openapi_operation_request_body

    from typing import Optional
    from pydantic.fields import ModelField
    from typing import Dict
    from typing import Union
    from typing import Type
    from pydantic import BaseModel
    from enum import Enum
    from typing import Any

    def get_request_body_with_explode(*,
        body_field: Optional[ModelField],
        model_name_map: Dict[Union[Type[BaseModel], Type[Enum]], str],
    ) -> Optional[Dict[str, Any]]:
        original = orig_get_request_body(body_field=body_field, model_name_map=model_name_map)
        if not original:
            return original
        content = original.get("content", {})
        form_patch = content.get("application/x-www-form-urlencoded", {})
        if not form_patch:
            form_patch = content.get("multipart/form-data", {})
        if form_patch:

            form_patch["encoding"] = {
                "vals": {  # TODO: change to be more dynamic
                    "style": "form",
                    "explode": True
                }
            }

        return original

    fastapi.openapi.utils.get_openapi_operation_request_body = get_request_body_with_explode
```

You would monkeypatch that in the same sort of place you instantiate `FastAPI`, or just any time before you invoke the `app.openapi()` function.

This works great with an endpoint defined like:

```python
@router.post("/withlistparams")
async def with_list_params(vals: list[str] = Form()):
    raise NotImplementedError()
```

The annoying part in the spec is that you have to specify the "encoding" on a per-request-property basis. But you could use the given `body_field` to look up the properties in the schema that are lists and ensure all of those properties have the right `encoding` set on the operation.

Hope that helps!

**Comment by falkben** [comment]
Thanks @phillipuniverse, that is very helpful. The code I had was directly modifying the schema dictionary, so I think this would be an improvement, although I think I want to come up with a way to generalize it to all List/form params. Unfortunately, that typing information doesn't seem to be available in the `get_request_body_with_explode` function so I'm still thinking about this.

> If there are changes to code, it should be within Swagger UI, not in FastAPI or Pydantic (although perhaps FastAPI could make it easier to extend the encoding).

I'm curious why you say that. It seems that what SwaggerUI produces is fine, generally speaking -- it's just that FastAPI doesn't parse lists of params with CSV so it cannot use the default that SwaggerUI produces.

Since that is the case, I would think the fix would naturally go into FastAPI where it generates the openapi scheme. Since it knows that `List` params can only be correctly parsed with the "exploded" duplicated params style, fastapi should generate an openapi scheme that has the `explode: true` option on by default.

Re: easier extending into the encoding section, I did see some interesting options there. For example, I know that FastAPI can't easily convert complex, nested Pydantic models to Form fields. However, there is an option in openapi called "Complex Serialization in Form Data" ([here](https://swagger.io/docs/specification/describing-request-body/)) where a form param will parse a JSON payload. This could be a "natural" way to enable nested models in Forms.

**Comment by phillipuniverse** [comment]
> Unfortunately, that typing information doesn't seem to be available in the get_request_body_with_explode function so I'm still thinking about this.

☹️  bummer.

> I'm curious why you say that. It seems that what SwaggerUI produces is fine, generally speaking -- it's just that FastAPI doesn't parse lists of params with CSV so it cannot use the default that SwaggerUI produces.

@falkben whether or not FastAPI should or should not parse CSV values is a bit separate from this discussion about Swagger UI. 

> Since that is the case, I would think the fix would naturally go into FastAPI where it generates the openapi scheme. Since it knows that List params can only be correctly parsed with the "exploded" duplicated params style, fastapi should generate an openapi scheme that has the explode: true option on by default.

I have read a bit more into the OpenAPI spec and [opened a bug ticket in swagger-js](https://github.com/swagger-api/swagger-js/issues/2713). In the bug ticket I included a bit more information and relevant parts of the OpenAPI spec, specifically that if not provided, the `encoding` object should have a default `style` of `form` for form-based request bodies, which then means that the default value of `explode` becomes `true`. So as far as the OpenAPI specification is concerned, the spec that FastAPI generates (no `encoding` at all) is consistent with its behavior to not parse arrays in the CSV form. So no change to FastAPI should be required.

If you look at the bug ticket that should convince you that the default the SwaggerUI produces is inconsistent with the OpenAPI specification. Let me know if I'm misreading!

If I'm not misreading then we should close this ticket in FastAPI as effectively a duplicate of https://github.com/tiangolo/fastapi/issues/50#issuecomment-1267068112.

**Comment by falkben** [comment]
> the encoding object should have a default style of form for form-based request bodies

ah, that would explain some of my confusion when reading through the spec and then comparing to the behavior of SwaggerUI. The spec does state that form data should have a default of `explode: true`. I guess I just assumed you had to explicitly include the `style: "form"` but I agree it should be able to be determined by the content section.

I'm definitely in favor of this being fixed in SwaggerUI if they agree it's a bug.

Otherwise, I think it should be possible to look into the model when creating the schema to determine if it needs an encoding section. It probably won't be pretty, but I suspect there's a way to get the information we need through the `model_name_map`

**Comment by phillipuniverse** [comment]
@falkben I couldn't help myself, here's the fully dynamic version of my monkeypatch!!

```python
    import fastapi.openapi.utils
    orig_get_request_body = fastapi.openapi.utils.get_openapi_operation_request_body

    from typing import Optional
    from pydantic.fields import ModelField
    from pydantic.schema import field_schema
    from typing import Any, Dict, Union, Type
    from pydantic import BaseModel
    from fastapi.openapi.constants import REF_PREFIX

    from enum import Enum
    def get_request_body_with_explode(*,
        body_field: Optional[ModelField],
        model_name_map: Dict[Union[Type[BaseModel], Type[Enum]], str],
    ) -> Optional[Dict[str, Any]]:
        original = orig_get_request_body(body_field=body_field, model_name_map=model_name_map)
        if not original:
            return original
        content = original.get("content", {})
        form_patch = content.get("application/x-www-form-urlencoded", {})
        if not form_patch:
            form_patch = content.get("multipart/form-data", {})
        if form_patch:
            schema_reference, schemas, _ = field_schema(
                body_field, model_name_map=model_name_map, ref_prefix=REF_PREFIX
            )
            array_props = []
            for schema in schemas.values():  # type: Dict[str, Any]
                for prop, prop_schema in schema.get("properties", {}).items():
                    if prop_schema.get("type") == "array":
                        array_props.append(prop)

            form_patch["encoding"] = {prop: {"style": "form"} for prop in array_props} # could include "explode": True but not necessary in swagger-ui

        return original

    fastapi.openapi.utils.get_openapi_operation_request_body = get_request_body_with_explode
```

> I think it should be possible to look into the model when creating the schema to determine if it needs an encoding section

Yup, exactly! The key part to my above code is to get the actual OpenAPI schema from Pydantic, not just a reference to it. The schema comes back in the 2nd part of the tuple in the `field_schema` call:

```python
   schema_reference, schemas, _ = field_schema(
                body_field, model_name_map=model_name_map, ref_prefix=REF_PREFIX
            )
```

Once you've got that it's trivial to determine the array properties. Shouldn't be too bad to refactor that to a recursive version either.

Let me know if that works for you!

**Reply by darkvertex** [reply]
Broken after upgrading to Pydantic V2. 😢

**Reply by sammaphey** [reply]
I believe this would equivalently work with Pydantic V2:

```python
from fastapi._compat import (
    GenerateJsonSchema,
    JsonSchemaValue,
    ModelField,
)
from fastapi.types import ModelNameMap

def get_request_body_with_explode(
    *,
    body_field: ModelField | None,
    schema_generator: GenerateJsonSchema,
    model_name_map: ModelNameMap,
    field_mapping: dict[
        tuple[ModelField, Literal["validation", "serialization"]], JsonSchemaValue
    ],
    separate_input_output_schemas: bool = True,
) -> dict[str, Any] | None:
    original = orig_get_request_body(
        body_field=body_field,
        schema_generator=schema_generator,
        model_name_map=model_name_map,
        field_mapping=field_mapping,
        separate_input_output_schemas=separate_input_output_schemas,
    )
    if not original:
        return original
    content = original.get("content", {})
    if form_patch := (
        content.get("application/x-www-form-urlencoded") or content.get("multipart/form-data")
    ):
        array_props = []
        schema = body_field._type_adapter.json_schema()
        for prop, prop_schema in schema.get("properties", {}).items():
            if prop_schema.get("type") == "array":
                array_props.append(prop)
        form_patch["encoding"] = {prop: {"style": "form"} for prop in array_props}
    return original
```

**Reply by sammaphey** [reply]
For those interested this no longer works in fastapi version `0.119.0` because the `GenerateJsonSchema` is no longer imported in `_compat.py`. But `GenerateJsonSchema` is something that is supplied by `pydantic` anyways, and so shouldn't have been imported that way. Here is the new approach:

```python
from pydantic.json_schema import GenerateJsonSchema
from fastapi._compat import (
    JsonSchemaValue,
    ModelField,
)
from fastapi.types import ModelNameMap

def get_request_body_with_explode(
    *,
    body_field: ModelField | None,
    schema_generator: GenerateJsonSchema,
    model_name_map: ModelNameMap,
    field_mapping: dict[
        tuple[ModelField, Literal["validation", "serialization"]], JsonSchemaValue
    ],
    separate_input_output_schemas: bool = True,
) -> dict[str, Any] | None:
    original = orig_get_request_body(
        body_field=body_field,
        schema_generator=schema_generator,
        model_name_map=model_name_map,
        field_mapping=field_mapping,
        separate_input_output_schemas=separate_input_output_schemas,
    )
    if not original:
        return original
    content = original.get("content", {})
    if form_patch := (
        content.get("application/x-www-form-urlencoded") or content.get("multipart/form-data")
    ):
        array_props = []
        schema = body_field._type_adapter.json_schema()
        for prop, prop_schema in schema.get("properties", {}).items():
            if prop_schema.get("type") == "array":
                array_props.append(prop)
        form_patch["encoding"] = {prop: {"style": "form"} for prop in array_props}
    return original
```

**Comment by falkben** [comment]
This worked perfectly @phillipuniverse!

I modified it slightly to use the walrus operator (my first time using the walrus operator, in fact)

```python
    if form_patch := (
        content.get("application/x-www-form-urlencoded")
        or content.get("multipart/form-data")
    ):
```

**Comment by jagadeeshkoka** [comment]
Is this issue fixed in FastAPI side ? I am having this issue right now, 

```
{

    "detail": [
        {
            "loc": [
                "body",
                "listOfItems",
                "__root__"
            ],
            "msg": "value is not a valid list",
            "type": "type_error.list"
        }
    ]
}
```

```
class ListOfItems(BaseModel):
    __root__ : List[Item()]
class Item(BaseModel):
   itemId : ItemIdEnum = 'PLASTIC'
  itemName : str = 'Dolly'

```
I am getting above error

**Comment by sglazyrin** [comment]
hello guys!
Is it the same issue that I have ?
For the schema like this and for GET LIST, GET BY PK operation: 
```
class InvoiceHeaderSchema(DefaultSchema):
    retailer_id: int = pydantic.Field(default=None)
    line_items: typing.Optional[typing.List[InvoiceLineItemSchema]] = []
```
I am getting for line_items "[]" as an example of data that will be returned from BE
And I am describing it in the fastapi using following code:
```
        APIRouter().get(
            "/",
            response_model=typing.Union[InvoiceHeaderPaginatedResponse, list[InvoiceHeaderSchema]],  # type: ignore
        )(cls.get_list)
 ```

I am a bit confused....

I expect to get in the swagger example of the data that should be returned from backend..... for such array props.

**Comment by WilliamStam** [comment]
same issue my side. swagger picks up that it needs to be a list
![image](https://github.com/tiangolo/fastapi/assets/182800/f36f2e51-839a-4734-be6f-0c1b0b2d6c65)

and allows you to define it as an array

but the curl `  -d 'name=&username=&password=&permissions=a.b,a.c'` joins it up :( 

in postman however works correctly when sent like this. 
![image](https://github.com/tiangolo/fastapi/assets/182800/7ca9ad09-af63-4a33-859a-ed33933dadfa)

im sure its a bug in swagger. but the same thing happens in rapidocs as well. redoc doesnt want to play nice with me at the moment so cant test that. 

so since it is being submitted as a list... ["aaa,bbb"] it is valid at least so can probably do something like for each value in param split on comma and add it to a new list. strange that 2 documentation systems actualy get it right to ask for multiples but end up submitting it as 1

and since i ticked that box that i would try help around here when submitting my own discussion :P here goes my "solution" 

```
async def post_user(
    permissions: List[str] = Form(None)
) -> UserSaveResponse:
    if permissions:
        print(type(permissions))
        new_permissions = []
        for item in permissions:
            for t in item.split(","):
                if t:
                    new_permissions.append(t)
        print(new_permissions)

```

**Reply by calcuboy** [reply]
I meet the same problem. And I don't think handle it in backend code is helpful, for the Swagger UI is just a dev tool,
Our front site will work like the posman do.
It's just annoying.

**Comment by danielborges93** [comment]
Here I solved the problem by specifying openapi_extra values ​​directly in the route, for example:

```python
@router.post(
    path='/test',
    openapi_extra={
        'requestBody': {
            'content': {
                'multipart/form-data': {
                    'encoding': {
                        'ids': {
                            'style': 'form',
                            'explode': True,
                        },
                    },
                },
            },
        },
    },
)
async def test(
        file: UploadFile = File(),
        ids: list[UUID] = Form(),
):
```

I'm using FastAPI 0.115.5.

**Reply by viniciusCalcantara** [reply]
Nice!
